### DISCUSSION TITLE
Getting "NoneType" object is not callable on AsyncClient destruction

### DISCUSSION THREAD

**Discussion by parafoxia** [encode/httpx#1967] [Potential Issue]
Whenever I make a request as part of a script, I always get the following error:
```
Exception ignored in: <function AsyncClient.__del__ at 0x7efcfb42a830>
Traceback (most recent call last):
  File "/home/parafoxia/Programs/Projects/analytix/.venv/lib/python3.10/site-packages/httpx/_client.py", line 2012, in __del__
TypeError: 'NoneType' object is not callable
```
The actual request is handled perfectly, but then it errors on close every time. I've tried running it in an event loop and via `asyncio.run`, and the same issue occurs.

I'm using WSL (Debian 11) on Windows 10. CPython 3.10.

**Comment by lovelydinosaur** [comment]
That doesn't make much sense against the current version of `httpx`.

As you can see line 2012 isn't calling something that could (sensibly) be `None`...

https://github.com/encode/httpx/blob/3c5884eeaab92eb751771bda08ef23ac3b908491/httpx/_client.py#L1987-L2016

I'd suggest:

* Double check you've got the latest version installed.
* Include the full traceback if you're still seeing errors.
* Take a look at source file referenced in the error, and show me exactly what line of code the error is referencing.
* Include the *simplest possible* example that you can, which still replicates the issue.

**Reply by parafoxia** [reply]
I forgot to update this -- turns out it was caused by either the client session not being closed, or the async loop not being closed (as in either causes the same error). I'm pretty sure I was on the right version because I checked the source directly for the installed version in the lib and it was the same as above, but not 100% sure.

Refactoring stuff to run it all from a single `asyncio.run(main())` method solved it though.
